app/recipe/tests/test_recipe_api.py

```diff
@@ -0,0 +1,211 @@
+from django.test import TestCase
+from django.urls import reverse
+
+from rest_framework import status
+from rest_framework.test import APIClient
+
+from core.models import Recipe, Ingredient
+
+
+RECIPES_URL = reverse('recipe:recipe-list')
+
+
+def _create_recipe(self):
+    payload = {
+        'name': 'Pizza',
+        'description': 'Put it in the oven',
+        'ingredients': [
+            {'name': 'dough'},
+            {'name': 'cheese'},
+            {'name': 'tomato'}
+        ]
+    }
+    res = self.client.post(RECIPES_URL, data=payload, format='json')
+    return res
+
+
+def detail_url(recipe_id):
+    """Return recipe detail URL"""
+    return reverse('recipe:recipe-detail', args=[recipe_id])
+
+
+def search_url(name):
+    """Return recipe search URL"""
+    return reverse('recipe:recipe-list') + f'?name={name}'
+
+
+class RecipeApiTest(TestCase):
+    """Tests for the Recipe API"""
+
+    def setUp(self):
+        self.client = APIClient()
+
+    def test_create_recipe_with_no_ingredients(self):
+        """Test creating a recipe with no ingredients fails"""
+        payload = {
+            'name': 'Pizza',
+            'description': 'Put it in the oven'
+        }
+        res = self.client.post(RECIPES_URL, data=payload, format='json')
+
+        self.assertEqual(res.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_create_recipe_with_ingredients(self):
+        """Test creating a recipe with ingredients is successful"""
+        res = _create_recipe(self)
+
+        self.assertEqual(res.status_code, status.HTTP_201_CREATED)
+        recipe = Recipe.objects.get(id=res.data['id'])
+        self.assertEqual(getattr(recipe, 'name'), 'Pizza')
+        self.assertEqual(
+            getattr(recipe, 'description'), 'Put it in the oven'
+        )
+        ingredients = Ingredient.objects.filter(recipe=res.data['id'])
+        self.assertEqual(len(ingredients), 3)
+        for ingredient in ingredients:
+            self.assertIn(
+                getattr(ingredient, 'name'), 'dough cheese tomato'
```

Comment:
Why do you use `getattr` here?

Reply:
This was what was done in the Udemy Django course. Is there a better way of doing this?

Reply:
This is a matter of taste, but generally speaking, if you know that the object has a certain attribute, their is no need to use `getattr` and you can access it directly through `ingredient.name` (because ingredient always has the `name` attribute).

Reply:
Good call.

Reply:
Made that change as I prefer just calling `ingredient.name` too - it's neater.